PubSub/src/PubSubClient.php

```diff
@@ -807,13 +892,69 @@ private function subscriptionFactory($name, $topic = null, array $info = [])
     public function __debugInfo()
     {
         $debugInfo = [];
-        if ($this->connection) {
-            $debugInfo['connection'] = get_class($this->connection);
-        }
-
         $debugInfo['projectId'] = $this->projectId;
         $debugInfo['encode'] = $this->encode;
 
         return $debugInfo;
     }
+
+    /**
+     * Helper function that initializes the default config for handlwritten clients.
+     *
+     * @param array $config The client config passed on by the user.
+     * @param array $customAttrs The key/val pairs containing default values
+     * common for the handwritten clients.
+     *
+     * @return array The default config combined with the options passed on by the user
+     * and auth related config.
+     */
+    private function getDefaultClientConfig(array $config, array $customAttrs)
+    {
+        $defaultScope = $customAttrs['defaultScope'] ?? [];
+        $emulatorHost = $customAttrs['emulatorHost'] ?? '';
+        
+        $config += [
+            'scopes' => [$defaultScope],
+            'projectIdRequired' => true,
+            'hasEmulator' => (bool) $emulatorHost,
+            'emulatorHost' => $emulatorHost
+        ];
+
+        if (isset($customAttrs['libVersion'])) {
+            $config['libVersion'] = $customAttrs['libVersion'];
+        }
+
+        $config = $this->configureAuthentication($config);
+
+        return $config;
+    }
+
+    /**
+     * Helper function that prepares the GAPIC classes list to be passed in to the
+     * RequestHandler before it's instantiation.
+     * For example, for PubSub if the user has passed 'gapicPublisherClient' in the $config
+     * it will be used as a GAPIC class while sending the requests via the RequestHandler.
+     *
+     * Otherwise the default publisher GAPIC client will be used.
+     *
+     * @param array $config The client config.
+     * @param array $classConfigMap A key/value pair where key is the GAPIC class
+     *  and the value is the config key name.
+     */
+    private function getGapicsFromConfig(array $config, array $classConfigMap)
```

Comment:
As mentioned in the decleration of `self::GAPIC_KEYS`, this only used if we want the user to be able to override the GAPIC class used for an operation.

For example, if the user uses something like this:

```
$client = new PubSubClient([
  'gapicPublisherClient' => 'USER_DEFINED_CLASS'
]);
```

Then this method will make sure that the `USER_DEFINED_CLASS` is sent to the `RequestHandler` and not `PublisherGapicClient` class that we use otherwise.